rclpy/src/rclpy/client.cpp

```diff
@@ -80,22 +58,31 @@ client_create(
     client_ops.qos = *qos_profile;
   }
 
-  // Use smart pointer to make sure memory is free'd on error
-  auto deleter = [](rclpy_client_t * ptr) {_rclpy_destroy_client(ptr);};
-  auto cli = std::unique_ptr<rclpy_client_t, decltype(deleter)>(
-    static_cast<rclpy_client_t *>(PyMem_Malloc(sizeof(rclpy_client_t))),
-    deleter);
-  if (!cli) {
-    throw std::bad_alloc();
-  }
-  cli->client = rcl_get_zero_initialized_client();
-  cli->node = node;
+  // Create a client
+  rcl_client_ = std::shared_ptr<rcl_client_t>(
+    new rcl_client_t,
```

Comment:
@sloretz meta: why the mixure of allocation sources? I see `PyMem_Alloc`, `new`, and `allocator.allocate` across the code base.

Reply:
Why not use the smart pointer library to allocate the object? http://www.cplusplus.com/reference/memory/make_shared/

Reply:
>  why the mixure of allocation sources?

Laziness. `new` throws if allocation fails, so the code doesn't have to check PyMem_Malloc's return or cleanup the `shared_ptr` to prevent attempting to free it twice. Maybe a PyMem_Malloc wrapper that throws on failure would be convenient here.

> Why not use the smart pointer library to allocate the object? http://www.cplusplus.com/reference/memory/make_shared/

I didn't use `make_shared` because I don't think it allows a custom deleter.